netaddr.go

```diff
@@ -252,6 +252,75 @@ func (ip IP) IsMulticast() bool {
 	}
 }
 
+// Mask applies a CIDR mask of leading bits to IP, producing an IPPrefix
+// of the specified length. If IP is the zero value or bits is too large, a
+// zero-value IPPrefix is returned.
+func (ip IP) Mask(bits uint8) IPPrefix {
+	maxBits := uint8(32)
+	if ip.Is6() {
+		maxBits = 128
+	}
+
+	if bits > maxBits {
+		// Invalid mask size for both IPv4/6: nothing to do.
+		return IPPrefix{}
+	}
+
+	var b []byte
+	switch ip := ip.ipImpl.(type) {
+	case nil:
+		// Zero-value input produces zero-value output.
+		return IPPrefix{}
+	case v4Addr:
+		b = ip[:]
+	case v6Addr:
+		b = ip[:]
+	case v6AddrZone:
+		b = ip.v6Addr[:]
+	default:
+		panic("netaddr: unhandled ipImpl representation")
+	}
+
+	// Apply the mask specified by bits.
+	n := bits
+	for i := 0; i < len(b); i++ {
+		if n >= 8 {
+			b[i] &= 0xff
+			n -= 8
+			continue
+		}
+
+		b[i] = ^byte(0xff >> n)
+		n = 0
+	}
+
+	var out IP
+	switch ip.ipImpl.(type) {
+	case v4Addr:
+		var v4 v4Addr
+		copy(v4[:], b)
+		out = IP{ipImpl: v4}
+	case v6Addr:
+		var v6 v6Addr
+		copy(v6[:], b)
+		out = IP{ipImpl: v6}
+	case v6AddrZone:
+		// TODO: does it makes sense to apply zone in an IPPrefix?
+		var v6 v6AddrZone
+		copy(v6.v6Addr[:], b)
+		v6.zone = ip.Zone()
+		out = IP{ipImpl: v6}
+	default:
+		panic("netaddr: unhandled ipImpl representation")
+	}
+
+	// TODO: return bare IP instead?
```

Comment:
perhaps? MaskPrefix could then be a couple line wrapper that returns IPPrefix isntead to save the caller a few lines?

@danderson, @crawshaw?

Reply:
Off the top of my head, I can't think why I would want to mask an IP for reasons other than constructing a prefix. Maybe if I want to do some kind of enumeration of the network, or construct one IP from another?

Having Mask return a bare IP seems fine if someone has a use in mind for it. I can't think of one.

Reply:
I would generally agree you want the prefix if you are masking. But it's trivial to construct an  IPPrefix yourself since you already had to specify the input bits anyway:

```go
const bits = 64
masked := ip.Mask(bits)
p := netaddr.Prefix{IP: masked, Bits: bits}
```

Reply:
Similarly it's very easy to extract just the IP from the returned prefix:

```
masked := ip.Mask(bits).IP
```

Given there's no functionality difference and no obvious performance difference (I would expect an optimizing compiler to inline this function then remove the Prefix creation in the above example), I say go with the return type that's more common. In this case I suspect it's the whole IPPrefix.

Reply:
Ack, will do.